### ISSUE TITLE
Boxed combinators?

### ISSUE THREAD

**Issue by clayrat** [gallais/idris-tparsec#10]
Using combinators on `rec` seems fairly common, and it seems to always require defining helpers to help with type inference, e.g.:

```
andbox : All (Box (Parser' s) :-> Box (Parser' t) :-> Box (Parser' (s, t)))
andbox {s} {t} p q = map2 {a=Parser' s} {b=Parser' t} (\p, q => and p q) p q

nelistbox : All (Box (Parser' s) :-> Box (Parser' (NEList s)))
nelistbox {s} p = map {a=Parser' s} nelist p
```

Should we maybe provide a separate module for these out of box, or there's a more elegant solution?

**Comment by clayrat** [comment]
> Apparently an implementation cannot mention a universally quantified variable standing for a value. All the arguments have to be either types or constructors...

Yeah, that's a known limitation in current Idris. You can do it for named implementations:
```
[appbox] Applicative (\ a => Box (Parser' a) n) where
  ...
```
but that will require wrapping every usage of it in `using implementation [appbox]` or passing it via the `@{..}` syntax, which is not so elegant.

There is, however, [a light at the end of the tunnel..](https://github.com/edwinb/Blodwen/commit/47dcc5b98775b2c53fb6113ef5c07327ecf82fb1)

**Comment by gallais** [comment]
Wait. Does that mean that if we had a `BoxedParser : Nat -> Type -> Type` we would
be able to define an instance for `BoxedParser n`?

**Comment by gallais** [comment]
Ah. No. Nevermind. Plus it wouldn't fit in the `interface` anyway because
`Box`'s argument is a `Nat`-indexed Type family, not a mere Type...

**Comment by clayrat** [comment]
I don't think so, `BoxedParser` would evaluate to a lambda and cause the exact same error :( 

I've just discovered though, that you can in fact inline the helper - writing `map {a=Parser' _} nelist rec` inside a larger expression seems to work fine.

**Comment by gallais** [comment]
What if you write `map {a=Parser _ _} nelist rec` instead? Because if that's enough
then we can consider writing a module with specialised versions of `map` and `app`!

**Comment by clayrat** [comment]
Yeah, writing `map {a=Parser _ _ _ _} nelist rec` works! I'm using 4 params here since the project using this is still based on the `master` branch.

**Comment by gallais** [comment]
In the #36 PR I have added a bunch of combinators called some variants of `lift`
that take a function on parser and make it a function on boxed parsers provided
that the return type is a boxed parser.

Once I had this my life was a lot easier as I could propagate the `Box` wherever
I wanted without having to deal with the implicits Idris cannot infer when using
`map2`.
